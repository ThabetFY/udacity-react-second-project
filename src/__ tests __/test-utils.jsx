// test-utils.js
import { render as rtlRender } from "@testing-library/react";
import { createStore } from "redux";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";

import rootReducer from "../store/reducers";
import middleware from "../store/middleware";

function render(
  ui,
  {
    initialState,
    store = createStore(rootReducer, initialState, middleware),
    ...renderOptions
  } = {},
) {
  function Wrapper({ children }) {
    return (
      <Provider store={store}>
        <MemoryRouter {...renderOptions}>{children}</MemoryRouter>
      </Provider>
    );
  }
  return { store, ...rtlRender(ui, { wrapper: Wrapper, ...renderOptions }) };
}

// re-export everything
export * from "@testing-library/react";

// override render method
export { render };
