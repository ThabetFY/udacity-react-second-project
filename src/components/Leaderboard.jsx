import { connect } from "react-redux";
import PropTypes from "prop-types";

const Leaderboard = ({ users }) => {
  console.log(users);
  return (
    <table>
      <thead>
        <tr>
          <th>Users</th>
          <th>Answered</th>
          <th>Created</th>
        </tr>
      </thead>
      <tbody>
        {users.map((user, index) => (
          <tr key={index}>
            <td>{user.name}</td>
            <td>{Object.keys(user.answers).length}</td>
            <td>{user.questions.length}</td>
          </tr>
        ))}
      </tbody>
    </table>
  );
};

const mapStateToProps = ({ users }) => {
  const sortedUsers = Object.values(users).sort(
    (a, b) =>
      b.questions.length +
      Object.keys(b.answers).length -
      (a.questions.length + Object.keys(a.answers).length)
  );
  return {
    users: sortedUsers,
  };
};

Leaderboard.propTypes = {
  users: PropTypes.array,
};

const connectedLeaderboard = connect(mapStateToProps)(Leaderboard);

export default connectedLeaderboard;
