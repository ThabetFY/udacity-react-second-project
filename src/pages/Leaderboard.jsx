import { useSelector } from "react-redux";

import {
  Table,
  TableBody,
  TableCell,
  TableHead,
  TableHeader,
  TableRow,
} from "@/components/ui/table";
import {
  Card,
  CardContent,
  CardDescription,
  CardHeader,
  CardTitle,
} from "@/components/ui/card";

const Leaderboard = () => {
  const users = useSelector((state) => state.users);
  const sortedUsers = Object.values(users).sort(
    (a, b) =>
      b.questions.length +
      Object.keys(b.answers).length -
      (a.questions.length + Object.keys(a.answers).length),
  );
  return (
    <Card className="shadow-md rounded-lg w-1/2 ">
      <CardContent>
        <Table>
          <TableHeader>
            <TableRow>
              <TableHead>Users</TableHead>
              <TableHead>Answered</TableHead>
              <TableHead>Created</TableHead>
            </TableRow>
          </TableHeader>
          <TableBody>
            {sortedUsers.map((user) => (
              <TableRow key={user.id}>
                <TableCell className="font-medium">{user.name}</TableCell>
                <TableCell>{Object.keys(user.answers).length}</TableCell>
                <TableCell>{user.questions.length}</TableCell>
              </TableRow>
            ))}
          </TableBody>
        </Table>
      </CardContent>
    </Card>
  );
};

export default Leaderboard;
