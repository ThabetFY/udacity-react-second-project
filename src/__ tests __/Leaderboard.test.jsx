import { describe, test, expect } from "vitest";
import { waitFor } from "@testing-library/react";

import Leaderboard from "@/routes/Leaderboard";
import { render } from "./test-utils";

// Mock Redux store

describe("Leaderboard", () => {
  test("Leaderboard mounts properly", async () => {
    render(<Leaderboard />);
  });

  test("renders the leaderboard table", async () => {
    const initialState = {
      authedUser: "tylermcginnis",
      users: {
        user1: {
          id: "user1",
          name: "Alice",
          answers: { q1: "optionOne" },
          questions: ["q2", "q3"],
        },
        user2: {
          id: "user2",
          name: "Bob",
          answers: { q2: "optionTwo" },
          questions: [],
        },
        // Add more mock users if needed
      },
    };
    const { getByText } = render(<Leaderboard />, { initialState });
    await waitFor(() => {
      expect(getByText("Alice")).toBeInTheDocument();
      expect(getByText("Bob")).toBeInTheDocument();
    });
  });
});
