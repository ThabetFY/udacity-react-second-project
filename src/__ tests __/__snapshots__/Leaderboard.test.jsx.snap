// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Leaderboard > renders the leaderboard table 1`] = `
<DocumentFragment>
  <div
    class="border bg-card text-card-foreground shadow-md rounded-lg w-1/2"
  >
    <div
      class="p-6 pt-0"
    >
      <div
        class="relative w-full overflow-auto"
      >
        <table
          class="w-full caption-bottom text-sm"
        >
          <thead
            class="[&_tr]:border-b"
          >
            <tr
              class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
            >
              <th
                class="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0"
              >
                Users
              </th>
              <th
                class="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0"
              >
                Answered
              </th>
              <th
                class="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0"
              >
                Created
              </th>
            </tr>
          </thead>
          <tbody
            class="[&_tr:last-child]:border-0"
          >
            <tr
              class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
            >
              <td
                class="p-4 align-middle [&:has([role=checkbox])]:pr-0 font-medium"
              >
                Alice
              </td>
              <td
                class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
              >
                1
              </td>
              <td
                class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
              >
                2
              </td>
            </tr>
            <tr
              class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
            >
              <td
                class="p-4 align-middle [&:has([role=checkbox])]:pr-0 font-medium"
              >
                Bob
              </td>
              <td
                class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
              >
                1
              </td>
              <td
                class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
              >
                0
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
